File: 2015-elixir/lib/solutions/day_05.ex
defmodule Aoc.Solution.Day05 do
  @name "Day 5: Doesn't He Have Intern-Elves For This?"
  @behaviour Solution

  @impl Solution
  def get_name, do: @name

  @impl Solution
  def parse!(raw) do
    raw
    |> String.split()
  end

  @impl Solution
  def solve_first_part(strings) do
    strings
    |> Enum.map(fn s -> nice?(s) end)
    |> Enum.map(fn b ->
      if b, do: 1, else: 0
    end)
    |> Enum.sum()
  end

  @impl Solution
  def solve_second_part(strings) do
    strings
    |> Enum.map(fn s -> still_nice?(s) end)
    |> Enum.map(fn b ->
      if b, do: 1, else: 0
    end)
    |> Enum.sum()
  end

  defp nice?(subject) do
    Enum.all?([
      does_not_contain_the_strings(subject),
      contains_at_least_three_vowels(subject),
      contains_at_least_one_letter_pair(subject)
    ])
  end

  defp still_nice?(subject) do
    Enum.all?([
      contains_a_letter_pair_without_overlapping(subject),
      contains_one_letter_pair_which_repeats_with_exactly_one_letter_between_them(subject)
    ])
  end

  defp contains_at_least_three_vowels(subject) do
    r = ~r/[aeiou].*[aeiou].*[aeiou].*/
    String.match?(subject, r)
  end

  defp contains_at_least_one_letter_pair(subject) do
    r = ~r/(\w)\1/
    String.match?(subject, r)
  end

  defp does_not_contain_the_strings(subject) do
    r = ~r/ab|cd|pq|xy/
    not String.match?(subject, r)
  end

  defp contains_a_letter_pair_without_overlapping(subject) do
    r = ~r/(\w{2}).*\1/
    String.match?(subject, r)
  end

  defp contains_one_letter_pair_which_repeats_with_exactly_one_letter_between_them(subject) do
    r = ~r/(\w)\w\1/
    String.match?(subject, r)
  end
end

File: 2015-python/solutions/day_05.py
import re

from solutions import BaseSolution


class Solution(BaseSolution):
    input_file = "05.txt"

    def __str__(self):
        return "Day 5: Doesn't He Have Intern-Elves For This?"

    def solve(self, pi):
        return self._solve(pi)[0]

    def solve_again(self, pi):
        return self._solve(pi)[1]

    def parse_input(self, data):
        return data.strip()

    def _solve(self, pi):
        wl = pi.split()
        p1 = sum(
            not re.search(r"ab|cd|pq|xy.*", w)
            and any(w[i] == w[i + 1] for i in range(len(w) - 1))
            and len(re.findall(r"[aeiou]", w)) > 2
            for w in wl
        )
        p2 = sum(
            any(w.count(w[i : i + 2]) == 2 for i in range(len(w) - 1))
            and any(w[i] == w[i + 2] for i in range(len(w) - 2))
            for w in wl
        )
        return p1, p2


if __name__ == "__main__":
    solution = Solution()
    solution.show_results()
